LiS/man/man4/testmod.4.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) $Id: testmod.4.man,v 0.9.2.3 2008-04-29 08:33:19 brian Exp $
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified $Date: 2008-04-29 08:33:19 $ by $Author: brian $
.\"
.\" -------------------------------------------------------------------------
.\"
.\" $Log: testmod.4.man,v $
.\" Revision 0.9.2.3  2008-04-29 08:33:19  brian
.\" - update headers for Affero release
.\"
.\" Revision 0.9.2.2  2006/04/05 09:42:35  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database lis.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so lis.macros
.\"
.\"
.TH TESTMOD 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Modules"
.\"
.\"
.SH NAME
.B testmod
\- \fISTREAMS\fR test module
.\"
.\"
.SH SYNOPSIS
.PP
.HP 8
.BI "int " fd ;
.HP 8
.BI "int open(" \*(lq\fR/dev/nuls\fP\*(rq ", " \fR0\fP );
.HP 8
.BI "int ioctl(" fd ", " I_PUSH ", " \*(lqtestmod\*(rq );
.\"
.\"
.SH DESCRIPTION
.PP
.B testmod
is a
.I STREAMS
module that is pushed over any
.IR Stream ,
.IR STREAMS -based
.BR pipe (4)
or
.IR STREAMS -based
.BR fifo (4).
.PP
Aside from being a
.I STREAMS
module,
.BR testmod
performs no actions on the data path.  All messages passed from the
.I Stream head
toward the
.B testmod
module are passed on downstream from the
.B testmod
write-side
.B put
procedure (with
.BR putnext (9));
all messaes received from the driver, are passed on upstream from the read-side
.B put
procedure.  The
.B testmod
module has no write- or read-side
.B service
procedure.
.PP
.B testmod
does, however, respond to a number of specialized input-output controls that are dedicated to
performing functions in the kernel necessary for the execution of test cases on a
.IR Stream .
These input-output controls are described below:
.\"
.SS IOCTLS
.PP
The following input-output control command can be executed on the file descriptor
.I fd
after the
.B testmod
module has been pushed:
.\"
.TP
.B TM_IOC_HANGUP
Send a
.BR M_HANGUP (9)
message to the
.IR "Stream head" .
.IP ""
This input-output control command takes no argument.
It is used primarily by
.BR test-streams (8)
to test the reaction to system calls executed on a
.I Stream
that is hung up.
.\"
.TP
.B TM_IOC_RDERR
Send a two-byte
.BR M_ERROR (9)
message to the
.I Stream head
with the read error set to
.I arg
and the write error set to
.BR NOERROR .
.IP ""
This input-output control command takes a single argument that contains the error number to be
placed in the read error byte of the
.BR M_ERROR (9)
message.
It is used primarily by
.BR test-streams (8)
to test the reaction to system calls executed on a
.I Stream
that has a read error.
.\"
.TP
.B TM_IOC_WRERR
Send a two-byte
.BR M_ERROR (9)
message to the
.I Stream head
with the read error set to
.B NOERROR
and the write error set to
.IR arg .
.IP ""
This input-output control command takes a single argument that contains the error number to be
placed in the write error byte of the
.BR M_ERROR (9)
message.
It is used primarily by
.BR test-streams (8)
to test the reaction to system calls executed on a
.I Stream
that has a write error.
.\"
.TP
.B TM_IOC_RWERR
Send a one-byte
.BR M_ERROR (9)
message to the
.I Stream head
with the error set to
.IR arg .
.IP ""
This input-output control command takes a single argument that contains the error number to be
placed in the
.BR M_ERROR (9)
message.
It is used primarily by
.BR test-streams (8)
to test the reaction to system calls executed on a
.I Stream
that has an error.
.\"
.\"
.SH USAGE
.PP
The primary purpose of
.B testmod
is as a test module for use with the
.BR test-streams (8)
conformance test program.  The other purpose of
.B testmod
is as an example of a
.I STREAMS
pushable module.
.\"
.\"
.SH NOTICES
.PP
.B testmod
is implementation-specific to the
.B @PACKAGE_TITLE@
package.
.\"
.\"
.SH EXAMPLES
.PP
The source code for
.B testmod
serves as an example for a
.I STREAMS
pushable module.
.\"
.\"
.SH MODULES
.PP

NEWS
OpenSS7 Master Package -- history of user visible changes.  2008-05-03
$Id: NEWS,v 0.9.2.53 2008-05-03 21:22:25 brian Exp $
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2000  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
OpenSS7 Master Package package.

 Major changes for release openss7-0.9.2.G
------------------------------------------

This is the sixth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Documented `lib32gcc1' problem on Ubuntu.

   - Noted problem running under XEN kernels.  XEN kernels are not yet
     supported.

   - License upgrade to AGPL v3.

   - Cross-compile fixes (strap out AC_FUNC_REALLOC macro when
     cross-compiling).

   - Ability to strap out major documentation build and installation
     primarily for embedded targets.

   - Improvements to common build process for embedded and
     cross-compile targets.

   - Support for flex 2.5.33  in maintainer mode.

   - Added the following sub-packages to the master build (and relase):

         - `strx25-0.9.2.1'


_This is a public stable production grade release of the package: it
deprecates previous releases.  Please upgrade to the current release
before reporting bugs._

As with other OpenSS7 releases, this release configures, compiles,
installs and builds RPMs and DEBs for a wide range of Linux 2.4 and 2.6
RPM- and DPKG-based distributions, and can be used on production
kernels without patching or recompiling the kernel.

This package is publicly released under the `GNU Affero General Public
License Version 3'.  The release is available as an `autoconf' tarball,
SRPM, DSC, and set of binary RPMs and DEBs.  See the downloads page
(http://www.openss7.org/download.html) for the `autoconf' tarballs,
SRPMs and DSCs.  For tarballs, SRPMs, DSCs and binary RPMs and DEBs,
see the openss7 package page (http://www.openss7.org/openss7_pkg.html).

See `http://www.openss7.org/codefiles/openss7-0.9.2.G/ChangeLog' and
`http://www.openss7.org/codefiles/openss7-0.9.2.G/NEWS' in the release
for more information.  Also, see the `openss7.pdf' manual in the
release (also in html `http://www.openss7.org/openss7_manual.html').

 Major changes for release openss7-0.9.2.F
------------------------------------------

This is the fifth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.


LiS/drivers/str/testmod.c
/*****************************************************************************

 @(#) $RCSfile: testmod.c,v $ $Name:  $($Revision: 0.9.2.14 $) $Date: 2008-04-29 08:33:11 $

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date: 2008-04-29 08:33:11 $ by $Author: brian $

 -----------------------------------------------------------------------------

 $Log: testmod.c,v $
 Revision 0.9.2.14  2008-04-29 08:33:11  brian
 - update headers for Affero release

 Revision 0.9.2.13  2007/08/15 04:57:59  brian
 - GPLv3 updates

 Revision 0.9.2.12  2007/08/14 10:46:57  brian
 - GPLv3 header update

 Revision 0.9.2.11  2007/07/14 01:32:56  brian
 - make license explicit, add documentation

 Revision 0.9.2.10  2007/04/01 12:22:00  brian
 - corrections to module loading

 Revision 0.9.2.9  2007/03/25 18:58:11  brian
 - changes to support 2.6.20-1.2307.fc5 kernel

 Revision 0.9.2.8  2007/03/25 00:51:03  brian
 - synchronization updates

 Revision 0.9.2.7  2006/12/18 09:50:46  brian
 - updated headers for release

 Revision 0.9.2.6  2006/10/27 22:38:54  brian
 - changes for 2.6.18 build

 Revision 0.9.2.5  2005/12/20 15:11:41  brian
 - result of SMP kernel testing for LiS

 Revision 0.9.2.4  2005/12/19 03:22:18  brian
 - wend for simple _RP

 Revision 0.9.2.3  2005/12/18 06:37:53  brian
 - corrections for binary compatiblity

 Revision 0.9.2.2  2005/12/12 12:26:27  brian
 - changes for warningless compile on gcc 4.0.2.

 Revision 0.9.2.1  2005/10/23 05:01:26  brian
 - test programs and modules for POSIX testing

 Revision 0.9.2.6  2005/10/14 12:26:49  brian
 - SC module and scls utility tested

 Revision 0.9.2.5  2005/10/13 10:58:44  brian
 - working up testing of sad(4) and sc(4)

 Revision 0.9.2.4  2005/10/07 09:34:24  brian
 - more testing and corrections

 Revision 0.9.2.3  2005/10/05 09:25:31  brian
 - poll tests, some noxious problem still with poll

 Revision 0.9.2.2  2005/09/27 23:34:25  brian
 - added test cases, tweaked straccess()

 Revision 0.9.2.1  2005/09/25 22:52:11  brian
 - added test module and continuing with testing

 *****************************************************************************/

#ident "@(#) $RCSfile: testmod.c,v $ $Name:  $($Revision: 0.9.2.14 $) $Date: 2008-04-29 08:33:11 $"

static char const ident[] =
    "$RCSfile: testmod.c,v $ $Name:  $($Revision: 0.9.2.14 $) $Date: 2008-04-29 08:33:11 $";

/*
 * This is TESTMOD a STREAMS test module that provides some specialized input-output controls meant
 * for testing purposes only.  It is pushed onto another driver (such as the nuls or echo driver) to
 * perform some more specialized functions.
 */

#include <linux/autoconf.h>
#include <linux/version.h>
#include <linux/module.h>
#include <linux/init.h>

#include <sys/kmem.h>
#include <sys/stream.h>
#include <sys/strconf.h>
#ifdef LFS
#include <sys/strsubr.h>
#endif
#include <sys/ddi.h>

#include <sys/testmod.h>

#ifdef LFS
#include "sys/config.h"
#endif

#define TESTMOD_DESCRIP		"UNIX SYSTEM V RELEASE 4.2 FAST STREAMS FOR LINUX"
#define TESTMOD_COPYRIGHT	"Copyright (c) 1997-2008 OpenSS7 Corporation.  All Rights Reserved."
#define TESTMOD_REVISION	"LfS $RCSfile: testmod.c,v $ $Name:  $($Revision: 0.9.2.14 $) $Date: 2008-04-29 08:33:11 $"
#define TESTMOD_DEVICE		"SVR 4.2 Test Module for STREAMS"
#define TESTMOD_CONTACT		"Brian Bidulock <bidulock@openss7.org>"
#define TESTMOD_LICENSE		"GPL"
#define TESTMOD_BANNER		TESTMOD_DESCRIP		"\n" \
				TESTMOD_COPYRIGHT	"\n" \
				TESTMOD_REVISION	"\n" \
				TESTMOD_DEVICE		"\n" \
				TESTMOD_CONTACT		"\n"
#define TESTMOD_SPLASH		TESTMOD_DEVICE		" - " \
				TESTMOD_REVISION	"\n"

#if defined LIS && defined MODULE
#define CONFIG_STREAMS_TESTMOD_MODULE MODULE
#endif

#ifdef CONFIG_STREAMS_TESTMOD_MODULE
MODULE_AUTHOR(TESTMOD_CONTACT);
MODULE_DESCRIPTION(TESTMOD_DESCRIP);
MODULE_SUPPORTED_DEVICE(TESTMOD_DEVICE);
MODULE_LICENSE(TESTMOD_LICENSE);
#if defined MODULE_ALIAS
MODULE_ALIAS("streams-testmod");
#endif
#endif

#ifndef CONFIG_STREAMS_TESTMOD_NAME
#ifdef LIS
#define CONFIG_STREAMS_TESTMOD_NAME TESTMOD__MOD_NAME
#else
#error "CONFIG_STREAMS_TESTMOD_NAME must be defined."
#endif
#endif
#ifndef CONFIG_STREAMS_TESTMOD_MODID
#ifdef LIS
#define CONFIG_STREAMS_TESTMOD_MODID TESTMOD__ID
#else
#error "CONFIG_STREAMS_TESTMOD_MODID must be defined."
#endif
#endif

modID_t modid = CONFIG_STREAMS_TESTMOD_MODID;

#ifndef module_param
MODULE_PARM(modid, "h");
#else
module_param(modid, ushort, 0444);
#endif
MODULE_PARM_DESC(modid, "Module ID for TESTMOD.");

#ifdef MODULE_ALIAS
#ifdef LFS
MODULE_ALIAS("streams-modid-" __stringify(CONFIG_STREAMS_TESTMOD_MODID));
MODULE_ALIAS("streams-module-testmod");
#endif
#endif

#ifdef LIS
#define STRMINPSZ   0
#define STRMAXPSZ   4096
#define STRHIGH	    5120
#define STRLOW	    1024
#endif

static struct module_info testmod_minfo = {
	.mi_idnum = CONFIG_STREAMS_TESTMOD_MODID,
	.mi_idname = CONFIG_STREAMS_TESTMOD_NAME,
	.mi_minpsz = STRMINPSZ,
	.mi_maxpsz = STRMAXPSZ,
	.mi_hiwat = STRHIGH,
	.mi_lowat = STRLOW,
};

static struct module_stat testmod_rstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));
static struct module_stat testmod_wstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));

/* 
 *  -------------------------------------------------------------------------
 *
 *  PUT routines
 *
 *  -------------------------------------------------------------------------
 */

#ifdef LIS
#define streamscall _RP
#endif

static streamscall int
testmod_rput(queue_t *q, mblk_t *mp)
{
	/* we don't queue so we don't need to worry about M_FLUSH */
	putnext(q, mp);
	return (0);
}

#ifdef LIS
union ioctypes {
	struct iocblk iocblk;
	struct copyreq copyreq;
	struct copyresp copyresp;
};

#define _printd(__x) while (0) { }
#define _ptrace(__x) while (0) { }

int
ctlmsg(unsigned char type)
{
	unsigned char mod = (type & ~QPCTL);

	return (((1 << mod) & ((1 << M_DATA) | (1 << M_PROTO) | (1 << (M_PCPROTO & ~QPCTL)))) == 0);
}

int
putnextctl2(queue_t *q, int type, int param1, int param2)
{
	mblk_t *mp;

	if (ctlmsg(type) && (mp = allocb(2, BPRI_HI))) {
		mp->b_datap->db_type = type;
		mp->b_wptr[0] = (unsigned char) param1;
		mp->b_wptr++;
		mp->b_wptr[1] = (unsigned char) param2;
		mp->b_wptr++;
		putnext(q, mp);
		return (1);
	}
	return (0);
}
#endif

static streamscall int
testmod_wput(queue_t *q, mblk_t *mp)
{
	union ioctypes *ioc = (typeof(ioc)) mp->b_rptr;
	int err = EINVAL;
	int rval = 0;

	/* we don't queue so we don't need to worry about M_FLUSH */
	switch (mp->b_datap->db_type) {
	case M_IOCTL:
		switch (ioc->iocblk.ioc_cmd) {
		case TM_IOC_HANGUP:
			/* It is convenient to by able to synthesize an M_HANGUP message upstream
			   so that the test program can test the response of a hung up stream to
			   various system calls and input-output commands.  This ioctl accomplishes 
			   that. */
			if (putnextctl(OTHERQ(q), M_HANGUP))
				goto ack;
			err = ENOSR;
			goto nak;
		case TM_IOC_RDERR:
		{
			int rwerr;

			rwerr = *(unsigned long *) mp->b_cont->b_rptr;
			_printd(("%s: error number is %d\n", __FUNCTION__, rwerr));
			/* Synthesize a M_ERROR message with a read error (equal to the arg) */
			if (putnextctl2(OTHERQ(q), M_ERROR, rwerr, NOERROR))
				goto ack;
			err = ENOSR;
			goto nak;
		}
		case TM_IOC_WRERR:
		{
			int rwerr;

			rwerr = *(unsigned long *) mp->b_cont->b_rptr;
			_printd(("%s: error number is %d\n", __FUNCTION__, rwerr));
			/* Synthesize a M_ERROR message with a write error (equal to the arg) */
			if (putnextctl2(OTHERQ(q), M_ERROR, NOERROR, rwerr))
				goto ack;
			err = ENOSR;
			goto nak;
		}
		case TM_IOC_RWERR:
		{
			int rwerr;

			rwerr = *(unsigned long *) mp->b_cont->b_rptr;
			_printd(("%s: error number is %d\n", __FUNCTION__, rwerr));
			/* Synthesize a M_ERROR message with an error (equal to the arg) */
			if (putnextctl1(OTHERQ(q), M_ERROR, rwerr))
				goto ack;
			err = ENOSR;
			goto nak;
		}
		case TM_IOC_PSIGNAL:
		{
			int signum;

			signum = *(unsigned long *) mp->b_cont->b_rptr;
			_printd(("%s: signal number is %d\n", __FUNCTION__, signum));
			/* Synthesize an M_SIG message with a signal (equal to the arg) */
			if (putnextctl1(OTHERQ(q), M_PCSIG, signum))
				goto ack;
			err = ENOSR;
			goto nak;
		}
		case TM_IOC_NSIGNAL:
		{
			int signum;

			signum = *(unsigned long *) mp->b_cont->b_rptr;
			_printd(("%s: signal number is %d\n", __FUNCTION__, signum));
			/* Synthesize an M_SIG message with a signal (equal to the arg) */
			if (putnextctl1(OTHERQ(q), M_SIG, signum))
				goto ack;
			err = ENOSR;
			goto nak;
		}
		case TM_IOC_COPYIN:
		{
			caddr_t uaddr;

			if (ioc->iocblk.ioc_count != TRANSPARENT || mp->b_cont == NULL
			    || mp->b_cont->b_datap->db_lim - mp->b_cont->b_datap->db_base <
			    FASTBUF) {
				_ptrace(("Error path taken!\n"));
				err = EINVAL;
				goto nak;
			}
#ifdef LFS
#if defined __LP64__
			if (ioc->iocblk.ioc_flag == IOC_ILP32)
				uaddr = (caddr_t) (unsigned long) (uint32_t)
				    *(unsigned long *) mp->b_cont->b_rptr;
			else
#endif				/* defined __LP64__ */
#endif				/* LFS */
				uaddr = (caddr_t) *(unsigned long *) mp->b_cont->b_rptr;
			mp->b_cont->b_wptr = mp->b_cont->b_rptr + FASTBUF;
			memset(mp->b_cont->b_rptr, 0xa5, FASTBUF);

			mp->b_datap->db_type = M_COPYIN;
			ioc->copyreq.cq_addr = uaddr;
			ioc->copyreq.cq_size = FASTBUF;
			ioc->copyreq.cq_flag = 0;
			ioc->copyreq.cq_private = NULL;
			goto reply;
		}
		case TM_IOC_COPYOUT:
		{
			caddr_t uaddr;

			if (ioc->iocblk.ioc_count != TRANSPARENT || mp->b_cont == NULL
			    || mp->b_cont->b_datap->db_lim - mp->b_cont->b_rptr < FASTBUF) {
				_ptrace(("Error path taken!\n"));
				err = EINVAL;
				goto nak;
			}
#ifdef LFS
#if defined __LP64__
			if (ioc->iocblk.ioc_flag == IOC_ILP32)
				uaddr = (caddr_t) (unsigned long) (uint32_t)
				    *(unsigned long *) mp->b_cont->b_rptr;
			else
#endif				/* defined __LP64__ */
#endif				/* LFS */
				uaddr = (caddr_t) *(unsigned long *) mp->b_cont->b_rptr;
			mp->b_cont->b_wptr = mp->b_cont->b_rptr + FASTBUF;
			memset(mp->b_cont->b_rptr, 0xa5, FASTBUF);

			mp->b_datap->db_type = M_COPYOUT;
			ioc->copyreq.cq_addr = uaddr;
			ioc->copyreq.cq_size = FASTBUF;
			ioc->copyreq.cq_flag = 0;
			ioc->copyreq.cq_private = NULL;
			goto reply;
		}
		case TM_IOC_IOCTL:
		{
			if (ioc->iocblk.ioc_count == TRANSPARENT) {
				_ptrace(("Error path taken!\n"));
				err = EINVAL;
				goto nak;
			}
			if (mp->b_cont == NULL
			    || mp->b_cont->b_wptr - mp->b_cont->b_rptr != ioc->iocblk.ioc_count) {
				_ptrace(("Error path taken!\n"));
				err = EINVAL;
				goto nak;
			}
			memset(mp->b_cont->b_rptr, 0xa5, ioc->iocblk.ioc_count);
			mp->b_datap->db_type = M_IOCACK;
			ioc->iocblk.ioc_rval = 0;
			ioc->iocblk.ioc_error = 0;
			goto reply;
		}
		case TM_IOC_COPYIO:
		{
			caddr_t uaddr;

			if (ioc->iocblk.ioc_count != TRANSPARENT || mp->b_cont == NULL
			    || mp->b_cont->b_datap->db_lim - mp->b_cont->b_datap->db_base <
			    FASTBUF) {
				_ptrace(("Error path taken!\n"));
				err = EINVAL;
				goto nak;
			}
#ifdef LFS
#if defined __LP64__
			if (ioc->iocblk.ioc_flag == IOC_ILP32)
				uaddr = (caddr_t) (unsigned long) (uint32_t)
				    *(unsigned long *) mp->b_cont->b_rptr;
			else
#endif				/* defined __LP64__ */
#endif				/* LFS */
				uaddr = (caddr_t) *(unsigned long *) mp->b_cont->b_rptr;
			mp->b_cont->b_wptr = mp->b_cont->b_rptr + FASTBUF;
			memset(mp->b_cont->b_rptr, 0xa5, FASTBUF);

			mp->b_datap->db_type = M_COPYIN;
			ioc->copyreq.cq_addr = uaddr;
			ioc->copyreq.cq_size = FASTBUF;
			ioc->copyreq.cq_flag = 0;
			ioc->copyreq.cq_private = (mblk_t *) uaddr;
			goto reply;
		}
		default:
			/* pass along anything we don't understand */
			break;
		}
		break;
	case M_IOCDATA:
		switch (ioc->copyresp.cp_cmd) {
		case TM_IOC_HANGUP:
		case TM_IOC_RDERR:
		case TM_IOC_WRERR:
		case TM_IOC_RWERR:
		case TM_IOC_PSIGNAL:
		case TM_IOC_NSIGNAL:
			if (ioc->copyresp.cp_rval != NULL)
				/* abort operations */
				goto free_it;
			break;
		case TM_IOC_COPYIN:
		{
			int i;

			if (ioc->copyresp.cp_rval != NULL)
				/* abort operations */
				goto free_it;
			if (!mp->b_cont || mp->b_cont->b_wptr - mp->b_cont->b_rptr != FASTBUF) {
				_ptrace(("Error path taken!\n"));
				err = EINVAL;
				goto nak;
			}
			for (i = 0; i < FASTBUF; i++)
				if ((unsigned char) mp->b_cont->b_rptr[i] != (unsigned char) 0xa5)
					break;
			if (i == FASTBUF) {
				_ptrace(("Error path taken!\n"));
				err = EINVAL;
				goto nak;
			}
			rval = 0;
			goto ack;
		}
		case TM_IOC_COPYOUT:
		{
			if (ioc->copyresp.cp_rval != NULL)
				/* abort operations */
				goto free_it;
			rval = 0;
			goto ack;
		}
		case TM_IOC_COPYIO:
		{
			int i;

			if (ioc->copyresp.cp_rval != NULL)
				/* abort operations */
				goto free_it;
			if (ioc->copyresp.cp_private != NULL) {
				if (!mp->b_cont
				    || mp->b_cont->b_wptr - mp->b_cont->b_rptr != FASTBUF) {
					_ptrace(("Error path taken!\n"));
					err = EINVAL;
					goto nak;
				}
				for (i = 0; i < FASTBUF; i++)
					if ((unsigned char) mp->b_cont->b_rptr[i] !=
					    (unsigned char) 0xa5)
						break;
				if (i == FASTBUF) {
					_ptrace(("Error path taken!\n"));
					err = EINVAL;
					goto nak;
				}
				memset(mp->b_cont->b_rptr, 0xa5, FASTBUF);

				mp->b_datap->db_type = M_COPYOUT;
				ioc->copyreq.cq_addr = (caddr_t) ioc->copyreq.cq_private;
				ioc->copyreq.cq_flag = 0;
				ioc->copyreq.cq_private = NULL;
				goto reply;
			}
			rval = 0;
			goto ack;
		}
		default:
			/* pass along anything we don't understand */
			break;
		}
		break;
	      ack:
		mp->b_datap->db_type = M_IOCACK;
		ioc->iocblk.ioc_count = 0;
		ioc->iocblk.ioc_rval = rval;
		ioc->iocblk.ioc_error = 0;
		goto reply;
	      nak:
		mp->b_datap->db_type = M_IOCNAK;
		ioc->iocblk.ioc_count = 0;
		ioc->iocblk.ioc_rval = -1;
		ioc->iocblk.ioc_error = err;
		goto reply;
	}
	putnext(q, mp);
	return (0);
      free_it:
	freemsg(mp);
	return (0);
      reply:
	qreply(q, mp);
	return (0);
}

/* 
 *  -------------------------------------------------------------------------
 *
 *  OPEN and CLOSE
 *
 *  -------------------------------------------------------------------------
 */
static streamscall int
testmod_open(queue_t *q, dev_t *devp, int oflag, int sflag, cred_t *crp)
{
	queue_t *wq;

	if (q->q_ptr != NULL) {
		return (0);	/* already open */
	}
	wq = WR(q);
	if (sflag == MODOPEN || wq->q_next != NULL) {
		/* inherit packet sizes of what we are pushed over */
		wq->q_minpsz = wq->q_next->q_minpsz;
		wq->q_maxpsz = wq->q_next->q_maxpsz;
		q->q_ptr = wq->q_ptr = q;	/* just set it to something */
		qprocson(q);
		return (0);
	}
	return (EIO);		/* can't be opened as driver */
}
static streamscall int
testmod_close(queue_t *q, int oflag, cred_t *crp)
{
	(void) oflag;
	(void) crp;
	if (!q->q_ptr)
		return (ENXIO);
	qprocsoff(q);
	q->q_ptr = WR(q)->q_ptr = NULL;
	return (0);
}

/* 
 *  -------------------------------------------------------------------------
 *
 *  Registration and initialization
 *
 *  -------------------------------------------------------------------------
 */
static struct qinit testmod_rinit = {
	.qi_putp = testmod_rput,
	.qi_qopen = testmod_open,
	.qi_qclose = testmod_close,
	.qi_minfo = &testmod_minfo,
	.qi_mstat = &testmod_rstat,
};

static struct qinit testmod_winit = {
	.qi_putp = testmod_wput,
	.qi_minfo = &testmod_minfo,
	.qi_mstat = &testmod_wstat,
};

static struct streamtab testmod_info = {
	.st_rdinit = &testmod_rinit,
	.st_wrinit = &testmod_winit,
};

#ifdef LFS
static struct fmodsw testmod_fmod = {
	.f_name = CONFIG_STREAMS_TESTMOD_NAME,
	.f_str = &testmod_info,
	.f_flag = D_MP,
	.f_kmod = THIS_MODULE,
};
#endif

#ifdef LFS
struct tm_ioctl {
	unsigned int cmd;
	void *opaque;
};

static struct tm_ioctl tm_map[] = {
	{.cmd = TM_IOC_HANGUP,}
	, {.cmd = TM_IOC_RDERR,}
	, {.cmd = TM_IOC_WRERR,}
	, {.cmd = TM_IOC_RWERR,}
	, {.cmd = TM_IOC_PSIGNAL,}
	, {.cmd = TM_IOC_NSIGNAL,}
	, {.cmd = TM_IOC_IOCTL,}
	, {.cmd = TM_IOC_COPYIN,}
	, {.cmd = TM_IOC_COPYOUT,}
	, {.cmd = TM_IOC_COPYIO,}
	, {.cmd = 0,}
};

static void
tm_unregister_ioctl32(void)
{
	struct tm_ioctl *i;

	for (i = tm_map; i->cmd != 0; i++)
		if (i->opaque != NULL)
			unregister_ioctl32(i->opaque);
}
static int
tm_register_ioctl32(void)
{
	struct tm_ioctl *i;

	for (i = tm_map; i->cmd != 0; i++) {
		if ((i->opaque = register_ioctl32(i->cmd)) == NULL) {
			tm_unregister_ioctl32();
			return (-ENOMEM);
		}
	}
	return (0);
}
#endif				/* LFS */

#ifdef CONFIG_STREAMS_TESTMOD_MODULE
static
#endif
int __init
testmod_init(void)
{
	int err;

#ifdef CONFIG_STREAMS_TESTMOD_MODULE
	printk(KERN_INFO TESTMOD_BANNER);
#else
	printk(KERN_INFO TESTMOD_SPLASH);
#endif
	testmod_minfo.mi_idnum = modid;
#ifdef LFS
	if ((err = tm_register_ioctl32()) < 0)
		return (err);
	if ((err = register_strmod(&testmod_fmod)) < 0) {
		tm_unregister_ioctl32();
		return (err);
	}
#endif
#ifdef LIS
	if ((err = lis_register_strmod(&testmod_info, CONFIG_STREAMS_TESTMOD_NAME)) < 0)
		return (err);
#endif
	if (modid == 0 && err > 0)
		modid = err;
#ifdef LIS
	if ((err = lis_register_module_qlock_option(modid, LIS_QLOCK_NONE)) < 0) {
		lis_unregister_strmod(&testmod_info);
		return (err);
	}
#endif
	return (0);
};

#ifdef CONFIG_STREAMS_TESTMOD_MODULE
static
#endif
void __exit
testmod_exit(void)
{
#ifdef LFS
	unregister_strmod(&testmod_fmod);
	tm_unregister_ioctl32();
#endif
#ifdef LIS
	lis_unregister_strmod(&testmod_info);
#endif
};

#ifdef CONFIG_STREAMS_TESTMOD_MODULE
module_init(testmod_init);
module_exit(testmod_exit);
#endif
